oio/common/service_resolver.py

```diff
@@ -0,0 +1,72 @@
+# Copyright (C) 2015-2019 OpenIO SAS, as part of OpenIO SDS
+#
+# This library is free software; you can redistribute it and/or
+# modify it under the terms of the GNU Lesser General Public
+# License as published by the Free Software Foundation; either
+# version 3.0 of the License, or (at your option) any later version.
+#
+# This library is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this library.
+
+
+from urlparse import urlparse, urlunparse
+
+from oio.common.green import time
+from oio.common.logger import get_logger
+from oio.conscience.client import ConscienceClient
+
+
+class ServiceResolver(object):
+    """A caching client to rawx services."""
+
+    def __init__(self, conf, max_age=60, logger=None):
+        self.conf = conf
+        self.logger = logger or get_logger(conf)
+        self.conscience_client = ConscienceClient(self.conf, logger=logger)
+
+        self.max_age = max_age
+        self._cache = dict()
+
+    def resolve_service_id(self, service_type, service_id,
+                           check_format=True):
+        """
+        :returns: Service address corresponding to the service ID
+        """
+        if check_format:
+            url = "http://" + service_id
+            parsed = urlparse(url)
+            if parsed.port is not None:
+                return service_id
+
+        cached_service_id = self._cache.get(service_id)
+        if cached_service_id \
+                and time.time() - cached_service_id['mtime'] < self.max_age:
+            return cached_service_id['addr']
+        result = self.conscience_client.resolve(
+            srv_type=service_type, service_id=service_id)
+        service_addr = result['addr']
+        self._cache[service_id] = {'addr': service_addr,
+                                   'mtime': time.time()}
+        return service_addr
+
+    def resolve_url(self, service_type, url):
+        """
+        :returns: Resolved URL of a service using a service ID
+        """
+        # FIXME(mb): some tests don't put scheme, should fix tests
+        if not url.startswith('http://'):
+            url = "http://" + url
+
+        parsed = urlparse(url)
+        if parsed.port is not None:
+            return url
```

Comment:
You've removed the `parsed.port == 80` it was your intention?

Reply:
Yes, it's intentional. There is no reason to make a difference for port 80.